kotlin/workflow-testing/src/main/java/com/squareup/workflow/testing/RenderTester.kt

```diff
@@ -0,0 +1,159 @@
+import com.squareup.workflow.Worker
+import com.squareup.workflow.Workflow
+import com.squareup.workflow.WorkflowAction
+import kotlin.reflect.KClass
+
+/**
+ * Create a [RenderTester] to unit test an individual render pass of this workflow.
+ *
+ * See [RenderTester] for usage documentation.
+ */
+@Suppress("UNCHECKED_CAST")
+fun <PropsT, OutputT : Any, RenderingT> Workflow<PropsT, OutputT, RenderingT>.renderTester(
+  props: PropsT
+): RenderTester<PropsT, *, OutputT, RenderingT> {
+  val statefulWorkflow = asStatefulWorkflow() as StatefulWorkflow<PropsT, Any?, OutputT, RenderingT>
+  return statefulWorkflow.renderTester(
+      props = props,
+      initialState = statefulWorkflow.initialState(props, null)
+  ) as RenderTester<PropsT, Nothing, OutputT, RenderingT>
+}
+
+/**
+ * Create a [RenderTester] to unit test an individual render pass of this workflow.
+ *
+ * See [RenderTester] for usage documentation.
+ */
+// @formatter:off
+fun <PropsT, StateT, OutputT : Any, RenderingT>
+    StatefulWorkflow<PropsT, StateT, OutputT, RenderingT>.renderTester(
+      props: PropsT,
+      initialState: StateT
+    ): RenderTester<PropsT, StateT, OutputT, RenderingT> =
+// @formatter:on
+  RealRenderTester(this, props, initialState)
+
+/**
+ * The props must be specified, the initial state may be specified, and then all child workflows
+ * and workers that are expected to run must be specified with [expectWorkflow] and [expectWorker]
+ * calls. Then call [render] and perform any assertions on the rendering. An event may also be sent
+ * to the rendering if no workflows or workers emitted an output. Lastly, the [RenderTestResult]
+ * returned by `render` may be used to assert on the [WorkflowAction]s processed to handle events or
+ * outputs by calling [verifyAction][RenderTestResult.verifyAction] or
+ * [verifyActionResult][RenderTestResult.verifyActionResult].
+ *
+ * - All workflows and workers that are rendered/ran by this workflow must be specified.
+ * - It is an error if more than one workflow or worker specifies an output.
+ * - It is a test failure if any workflows or workers that were expected were not ran.
+ * - It is a test failure if the workflow tried to run any workflows or workers that were not
+ *   expected.
+ * - It is a test failure if no workflow or workflow emitted an output, no rendering event was
+ *   invoked, and any of the action verification methods on [RenderTestResult] is called.
+ *
+ * ## Usage
+ *
+ * ```
+ * workflow
+ *   .renderTester(props = MyProps(…))
+ *   .expectWorker(
+ *     matchesWhen = { it is SubmitLoginWorker },
+ *     key = "signin",
+ *     output = EmittedOutput(LoginResponse(success = true))
+ *   )
+ *   .expectWorkflow(
+ *     workflowType = ChildWorkflow::class,
+ *     key = "child",
+ *     assertProps = { assertThat(it.email).isEqualTo("test@foo.com") },
+ *     rendering = ChildRendering("message")
+ *   )
+ *   .render { rendering ->
+ *     assertThat(rendering.text).isEqualTo("foo")
```

Comment:
Perhaps here could add a couple of (commented out?) lines demonstrating sending an event or having a child emit.

Reply:
A child does emit, on line 79.

Reply:
Added another example demonstrating this.